contracts/governance/extensions/GovernorCountingFractional.sol

```diff
@@ -0,0 +1,201 @@
+
+    error GovernorInvalidParamsFormat(address voter);
+    error GovernorUsedVotesExceedRemainingWeight(address voter, uint256 usedVotes, uint256 remainingWeight);
+
+    /**
+     * @dev See {IGovernor-COUNTING_MODE}.
+     */
+    // solhint-disable-next-line func-name-mixedcase
+    function COUNTING_MODE() public pure virtual override returns (string memory) {
+        return "support=bravo&quorum=for,abstain&params=fractional";
+    }
+
+    /**
+     * @dev See {IGovernor-hasVoted}.
+     */
+    function hasVoted(uint256 proposalId, address account) public view virtual override returns (bool) {
+        return voteWeightCast(proposalId, account) > 0;
+    }
+
+    /**
+     * @dev Get the number of votes cast thus far on proposal `proposalId` by account `account`. Useful for
+     * integrations that allow delegates to cast rolling, partial votes.
+     */
+    function voteWeightCast(uint256 proposalId, address account) public view virtual returns (uint256) {
+        return _proposalVotes[proposalId].usedVotes[account];
+    }
+
+    /**
+     * @dev Accessor to the internal vote counts.
+     */
+    function proposalVotes(
+        uint256 proposalId
+    ) public view virtual returns (uint256 againstVotes, uint256 forVotes, uint256 abstainVotes) {
+        ProposalVote storage proposalVote = _proposalVotes[proposalId];
+        return (proposalVote.againstVotes, proposalVote.forVotes, proposalVote.abstainVotes);
+    }
+
+    /**
+     * @dev See {Governor-_quorumReached}.
+     */
+    function _quorumReached(uint256 proposalId) internal view virtual override returns (bool) {
+        ProposalVote storage proposalVote = _proposalVotes[proposalId];
+        return quorum(proposalSnapshot(proposalId)) <= proposalVote.forVotes + proposalVote.abstainVotes;
+    }
+
+    /**
+     * @dev See {Governor-_voteSucceeded}. In this module, forVotes must be > againstVotes.
+     */
+    function _voteSucceeded(uint256 proposalId) internal view virtual override returns (bool) {
+        ProposalVote storage proposalVote = _proposalVotes[proposalId];
+        return proposalVote.forVotes > proposalVote.againstVotes;
+    }
+
+    /**
+     * @notice See {Governor-_countVote}.
+     *
+     * @dev Function that records the delegate's votes.
+     *
+     * If the `params` bytes parameter is empty, then this module behaves identically to {GovernorCountingSimple}.
+     * That is, it assigns the remaining weight of the delegate to the `support` parameter, which follows the
+     * `VoteType` enum from Governor Bravo (as defined in {GovernorCountingSimple}).
+     *
+     * If the `params` bytes parameter is not zero, then it _must_ be tree packed uint128s, totaling 48 bytes,
+     * representing the weight the delegate assigns to Against, For, and Abstain respectively. This format can be
+     * produced using:
+     *
+     * `abi.encodePacked(uint128(againstVotes), uint128(forVotes), uint128(abstainVotes))`
```

Comment:
I'd challenge this format a bit. It essentially means that any casted vote with values above `type(uint128).max` will need to vote twice. These may be handled on the app side, but are highly unlikely and most likely that justifies this strategy in the first place.

Shall we mention that in the docs? 2**128 = 3.4e38 and most tokens might have supply of < 1e15 with 18 decimals totalling e23. I think there's room enough to not be an issue but this is the kind of things Code4rena auditors look for 👀

Reply:
This format was "decided" by Scopelift, and is supported by Tally. If we change it, we break compatibility.

Reply:
Agree. What about?

```suggestion
     * If the `params` bytes parameter is not zero, then it _must_ be tree packed uint128s, totaling 48 bytes,
     * representing the weight the delegate assigns to Against, For, and Abstain respectively. This format can be
     * produced using:
     *
     * `abi.encodePacked(uint128(againstVotes), uint128(forVotes), uint128(abstainVotes))`
     *
     * NOTE: Consider the number of votes is restricted to 128 bits. Depending on how many decimals the underlying token
     * has, a single voter may require to split their vote into multiple transactions. For precision higher than
     * ~30 decimals, large token holders may require an exponentially large number of transactions to cast their vote.
```